src/retry-strategy/exponential-backoff-retry-strategy.ts

```diff
@@ -0,0 +1,196 @@
+import { NetworkRequest } from '../request';
+import { BaseSubscription } from '../subscription/base-subscription';
+import { ErrorResponse, NetworkError } from '../base-client';
+import { UnauthenticatedRetryStrategy } from './unauthenticated-retry-strategy';
+import { EmptyLogger, Logger } from '../logger';
+import { DoNotRetry, Retry, RetryStrategy, RetryStrategyResult } from './retry-strategy';
+
+export interface ExponentialBackoffRetryStrategyOptions {
+    tokenFetchingRetryStrategy?: RetryStrategy, //Retry strategy that checks for expired token and fetches it
+    retryUnsafeRequests?: boolean, //Elements doesn't allow unsafe requests to be retried, external calls to non-elements APIs might require it (for token providers, for instance)
+    limit?: number, //Max number of retries, -1 if unlimited
+    maxBackoffMillis?: number, //Maximum length for backoff
+    defaultBackoffMillis?: number, //Initial backoff we start from
+    logger?: Logger //A lumberjack
+}
+
+export class ExponentialBackoffRetryStrategy implements RetryStrategy {
+
+    private logger: Logger;
+    private tokenFetchingRetryStrategy: RetryStrategy;
+    private retryUnsafeRequests: boolean;
+    private limit: number;
+    private maxBackoffMillis: number;
+    private defaultBackoffMillis: number;
+
+    private retryCount: number = 0;
+    private currentBackoffMillis: number;
+    private pendingTimeout: number = 0;  
+    
+    constructor(private options: ExponentialBackoffRetryStrategyOptions){        
+        this.logger = this.options.logger || new EmptyLogger();
+        this.tokenFetchingRetryStrategy = this.options.tokenFetchingRetryStrategy || new UnauthenticatedRetryStrategy(this.logger);
+        this.retryUnsafeRequests = this.options.retryUnsafeRequests || false;
+
+        if(this.options.limit != null && this.options.limit != undefined){
+             this.limit = this.options.limit;
+        }
+        else{
+            this.limit = -1;
+        }
+        this.maxBackoffMillis = this.options.maxBackoffMillis || 5000;
+        this.defaultBackoffMillis = this.options.defaultBackoffMillis || 1000;
+        this.currentBackoffMillis = this.defaultBackoffMillis;
+    }
+
+    executeRequest<T>( 
+        error: any,
+        request: NetworkRequest<T>) {
+
+            if(!error){
+                return request().catch(error => {
+                    return this.executeRequest(error, request)
+                });
+            }
+
+            else{
+                return this.resolveError(error).then( () => {
+                    return this.executeRequest(null, request);
+                });
+            }
+    }
+    
+    executeSubscription(
+        error: any,
```

Comment:
Passing errors into methods signals that the code does not compose well. Why do you have to pass errors up the stack, instead of passing them down via promises/subscription callbacks?